providers/supermicro/supermicro.go

```diff
@@ -0,0 +1,343 @@
+	"github.com/pkg/errors"
+
+	bmclibconsts "github.com/bmc-toolbox/bmclib/v2/constants"
+	bmclibErrs "github.com/bmc-toolbox/bmclib/v2/errors"
+)
+
+const (
+	// ProviderName for the provider Supermicro implementation
+	ProviderName = "supermicro"
+	// ProviderProtocol for the provider supermicro implementation
+	ProviderProtocol = "vendorapi"
+)
+
+var (
+	// Features implemented
+	Features = registrar.Features{
+		providers.FeatureScreenshot,
+	}
+)
+
+// supports
+// SYS-5019C-MR
+// SYS-510T-MR
+// 6029P-E1CR12L
+
+type Config struct {
+	HttpClient           *http.Client
+	Port                 string
+	httpClientSetupFuncs []func(*http.Client)
+}
+
+// Option for setting optional Client values
+type Option func(*Config)
+
+func WithHttpClient(httpClient *http.Client) Option {
+	return func(c *Config) {
+		c.HttpClient = httpClient
+	}
+}
+
+// WithSecureTLS returns an option that enables secure TLS with an optional cert pool.
+func WithSecureTLS(rootCAs *x509.CertPool) Option {
+	return func(c *Config) {
+		c.httpClientSetupFuncs = append(c.httpClientSetupFuncs, httpclient.SecureTLSOption(rootCAs))
+	}
+}
+
+func WithPort(port string) Option {
+	return func(c *Config) {
+		c.Port = port
+	}
+}
+
+// Connection details
+type Client struct {
+	client    *http.Client
+	host      string
+	user      string
+	pass      string
+	port      string
+	csrfToken string
+	log       logr.Logger
+}
+
+// New returns connection with a Supermicro client initialized
+func NewClient(host, user, pass string, log logr.Logger, opts ...Option) *Client {
+	if !strings.HasPrefix(host, "https://") && !strings.HasPrefix(host, "http://") {
+		host = "https://" + host
+	}
+
+	defaultConfig := &Config{
+		Port: "443",
+	}
+
+	for _, opt := range opts {
+		opt(defaultConfig)
+	}
+
+	return &Client{
+		host:   host,
+		user:   user,
+		pass:   pass,
+		port:   defaultConfig.Port,
+		client: httpclient.Build(defaultConfig.httpClientSetupFuncs...),
+		log:    log,
+	}
+}
+
+// Open a connection to a Supermicro BMC using the vendor API.
+func (c *Client) Open(ctx context.Context) (err error) {
+	data := fmt.Sprintf(
+		"name=%s&pwd=%s&check=00",
+		base64.StdEncoding.EncodeToString([]byte(c.user)),
+		base64.StdEncoding.EncodeToString([]byte(c.pass)),
+	)
+
+	headers := map[string]string{"Content-Type": "application/x-www-form-urlencoded"}
+
+	body, status, err := c.query(ctx, "cgi/login.cgi", http.MethodPost, bytes.NewBufferString(data), headers, 0)
+	if err != nil {
+		return errors.Wrap(bmclibErrs.ErrLoginFailed, err.Error())
+	}
+
+	if status != 200 {
+		return errors.Wrap(bmclibErrs.ErrLoginFailed, strconv.Itoa(status))
+	}
+
+	if !bytes.Contains(body, []byte(`url_redirect.cgi?url_name=mainmenu`)) &&
+		!bytes.Contains(body, []byte(`url_redirect.cgi?url_name=topmenu`)) {
+		return errors.Wrap(bmclibErrs.ErrLoginFailed, "unexpected response contents")
+	}
+
+	contentsTopMenu, status, err := c.query(ctx, "cgi/url_redirect.cgi?url_name=topmenu", http.MethodGet, nil, nil, 0)
```

Comment:
might be helpful to have a code comment about needing to first POST to login and then do a GET for the csrf token. Seems like something i'm going to forget pretty quickly 😬

Reply:
noted, will include that in the next set of changes